asynciterator.ts

```diff
@@ -777,9 +789,95 @@ export class IntegerIterator extends AsyncIterator<number> {
   }
 }
 
+/**
+ * A synchronous mapping function from one element to another.
+ * The iterator performing the mapping is passed as a second argument.
+ */
+export type MapFunction<S, D = S> = (item: S) => D | null;
+
+/**
+ An iterator that calls a synchronous mapping function
+ on every item from its source iterator.
+ @extends module:asynciterator.AsyncIterator
+*/
+export class MappingIterator<S, D = S> extends AsyncIterator<D> {
+  private readonly _destroySource: boolean;
+  protected _source: InternalSource<S>;
+
+  /**
+   * Applies the given mapping to the source iterator.
+   */
+  constructor(
+    _source: AsyncIterator<S>,
+    private _map: MapFunction<S, D>,
+    options: SourcedIteratorOptions = {}
+  ) {
+    super();
+    this._source = _validateSource(_source);
+    this._destroySource = options.destroySource !== false;
+
+    if (_source.done) {
+      this.close();
+    }
+    else {
+      this._source._destination = this;
+      this._source.on('end', destinationClose);
+      this._source.on('error', destinationEmitError);
+      this._source.on('readable', destinationSetReadable);
+      // If we are given a source that is already readable
+      // then we need to set the state of this iterable to readable
+      // also as there is no guarantee that the is no forthcoming
+      // readable event from the source
+      this.readable = this._source.readable;
+    }
+  }
+
+  read(): D | null {
+    // Do not read the source if the current iterator is ended
+    if (this.done)
+      return null;
+
+    // Do not read from the source if it is ended
+    // TODO: Possibly remove this after https://github.com/RubenVerborgh/AsyncIterator/pull/45 is merged
+    // the requirement for this to be removed is that the 'end' emit occurs synchronously when the ENDED
+    // state is set, which does not currently occur if endAsync is called.
+    if (this._source.done) {
+      this.close();
+      return null;
+    }
+
+    let item: D | S | null;
+    while ((item = this._source.read()) !== null) {
+      if ((item = this._map(item)) !== null)
+        return item;
+    }
+
+    // This will set readable to false on the current iterator
+    // if the source is no longer readable
+    this.readable = this._source.readable;
+    return null;
+  }
+
+  public close() {
+    if (this._destroySource)
+      this._source.destroy();
+    super.close();
+  }
+
+  /* Cleans up the source iterator and ends. */
+  protected _end(destroy: boolean) {
+    this._source.removeListener('end', destinationClose);
+    this._source.removeListener('error', destinationEmitError);
+    this._source.removeListener('readable', destinationSetReadable);
+    delete this._source._destination;
+    if (this._destroySource)
+      this._source.destroy();
```

Comment:
That's double compared to `close`; can it be removed there then (and thus, can the entire `close` implementation be removed)?

Reply:
Ah yes it can, the only thing that changes is that the destruction happens a tick later.

This will become moot with https://github.com/RubenVerborgh/AsyncIterator/pull/45 anyway which removes the `endAsync` that results in this next-tick behavior.

Reply:
I've made this change. The only thing I should flag is that in https://github.com/RubenVerborgh/AsyncIterator/pull/75/commits/b34fb6e71647fd6c58afe0734271864d6220461a I had to change from waiting a tick (where tests were failing) to doing a `beforeEach` which waits for the end event on the iterator that is being closed, and then will wait another tick before testing for destruction.

I think this should be ok, but is probably something that we can make even safer in #45 as noted in the comments in that PR.